### ISSUE TITLE
App isn't saving anything 

### ISSUE THREAD

**Issue by baseblade** [mikebryant/ac-nh-turnip-prices#375]
**Things to check**
- Have you forced a refresh in your browser (e.g. ctrl+F5 in Chrome/Firefox)?
  - using mobile app
- Have you made sure the first time buyer option is correct? If you're unsure, try it both ways
  - yes
- Have you time-travelled this week? Travelling backwards resets the prices, and therefore you must not put prices in from both before & after the time-travel
  - /no

**Describe the bug**
yesterday the app completely cleared everything you said of me having to delete all reset and now there is nothing saving when I'm putting the prices and all my statistics have gone

**To Reproduce**

**Permalink to your prices**
Please click the `Copy Permalink` button and paste it here

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**
 - Device: Google Pixel 3XL
 - OS: unsure
 - Browser chrome 
 - Version ?

**Additional context**

**Comment by Tekaoh** [comment]
This is happening to me too using Chrome on Windows 10. I lost my save data late last week but then it appeared to be saving once again. But I lost my save data again this morning. This has never happened before. Something must have changed

**Comment by murielg** [comment]
Happening to me as well, onMac 10.14.5 Chrome v81, and iOS 13.3 Safari .

**Comment by Hattoff** [comment]
I think this may have something to do with this statement in your scripts.js file in the update function:
```
  if (!window.populated_from_query) {
    updateLocalStorage(prices, first_buy, previous_pattern);
  }
```
If someone used a URL to generate a portion of their data, then any new input will never be saved. A simple refresh of the page doesn't reveal this problem, as the browser will cache data input into these fields. This becomes more evident, however, on mobile devices, as they don't seem to keep pages persistent, and often force-refresh the browser when users close or leave the application for an extended period of time. I believe this is why some users report this problem intermittently, as though it happens randomly.

To replicate this behavior on PC:

- Open this copied permalink:

[https://turnipprophet.io?prices=109.139.123.321.........&pattern=3](url)

- Edit some values.
- Hit F5 and note that the values stay in place.
- Now hit CTRL + F5 and notice that the query was re-loaded and previously entered data was never actually saved.

I believe by updating that if statement on the update function in the scripts.js file to something like the following:
```
  if (!window.populated_from_query) {
    updateLocalStorage(prices, first_buy, previous_pattern);
  } else {
    updateLocalStorage(prices, first_buy, previous_pattern);
    window.populated_from_query = false;
    window.location.href = window.location.origin;
  }
```
It should resolve the problem by identifying if a query was loaded from the URL, cache the data loaded, then refreshing the page to strip the query. Thus preventing the browser from force-refreshing the page and reloading the original query.

I cloned the repo and tested this solution and it appears to work, but the page flashes a bit, as the values loaded from the URL are filled in before the data is cached and the page is refreshed. I don't know how to submit this as a potential solution, and I think a better fix could be implemented, but I could give it a shot if needed. Just let me know? Otherwise, I hope this helps anyone working on this bug.

-Hattoff

**Comment by renefloor** [comment]
I also have this problem since this weekend. I don't remember clicking on a permalink, but I do have the page installed as a PWA on android.

**Comment by norweeg** [comment]
The PWA on my google pixel just staraight up forgot all my prices this week

**Comment by Tekaoh** [comment]
> If someone used a URL to generate a portion of their data, then any new input will never be saved.

I've had my data lost without having used a URL to generate any portion of my data. I have a shortcut on the bookmarks bar in Chrome on Windows that just links to https://turnipprophet.io/ and I input my data right on the page. I visit via the shortcut one day, enter my data. I use the shortcut again the next day, my data is gone.

Just happened again to me this morning. I had data saved from Sunday, Monday, and Tuesday. Today is Wednesday. I used my shortcut in Chrome like always, but all my data is gone this morning.

Makes me think that something else is going on that might not be consistent with your theory @Hattoff

**Comment by Hattoff** [comment]
>I visit via the shortcut one day, enter my data. I use the shortcut again the next day, my data is gone.

Thanks for the additional information @Tekaoh . There is a boolean value being set which prevents data saves. Could you verify that it hadn't been inadvertently been set to true? 

Click on your shortcut and go to the console (hit _CTRL_ + _SHIFT_ + _i_ and click on the Console tab) and type `window.populated_from_query`  and hit enter. It should come back as _False_ or _Undefined_. If it comes back true then the page has cached that boolean improperly and none of your data is going to save. 

I will keep looking into this and see what I can come up with.
Thanks!
-Hattoff

**Comment by Tekaoh** [comment]
> Could you verify that it hadn't been inadvertently been set to true?

![undefined](https://user-images.githubusercontent.com/45337851/82455300-cc29bc00-9a80-11ea-9c57-e2c20a5a85f7.PNG)

Undefined as of this morning anyways. If it was somehow set to true when I used the app yesterday, maybe everything was deleted yesterday. But say for example it was true yesterday, would that delete _all_ my save data or just what I entered yesterday?

I'm kind of scratching my head on this one... I started experiencing issues only about a week ago. Before that, everything seemed to work normally. But I'm looking through all the commits for the last week or so and I don't see anything that might have changed the way data is saved in any meaningful way. Super odd... Gotta dig deeper into them I guess?

**Comment by Hattoff** [comment]
@Tekaoh 
I think I might have something. Rapid refreshes seem to trigger the update function before the data can be properly populated. The user data is pulled directly from the dom using an event trigger when a value is changed. This can be a problem when initializing the page because if there is any delay in the loading of the document, the fields will not be properly populated from data in the localStorage before the first update function is called. The values get wiped out when these functions are called too early on an empty dom, producing null values:

`  const sell_prices = getSellPrices();
  const buy_price = parseInt(buy_input.val());
  const first_buy = getCheckedRadio(first_buy_radios) == 'true';
  const previous_pattern = parseInt(getCheckedRadio(previous_pattern_radios));
`

The only code that recently changed which could potentially cause this is merge #311 to the js/contributors.js script, which pulls the contributors of this code using GitHub's API. I am able to consistently wipe my data by semi-rapidly refreshing the page. By doing so I noticed a data-wipe occurs when the browser decides to re-pull this data, rather than use the cached information.

![image](https://user-images.githubusercontent.com/18622607/82467038-689e8f00-9a7e-11ea-8ebe-99fb4aa5448d.png)

I think the problem isn't particularly with this function, but rather, how the data is stored, cached, and updated. It is relying on the data to be populated by the dom, which just isn't fast enough in some cases. I will work on a fix for that, but I think in the meantime, if this can be verified as what has triggered this bug, it might be best to roll back those changes until a better data management system can be devised.

Let me know if you can verify and what you think?

-Hattoff

**Comment by maxlefou** [comment]
It happened to me this morning. And on top of that, i can't put back data because the app freezes completely when i try to

**Comment by mikebryant** [comment]
I believe this has been fixed by #385
